Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/StateSynchronization/CommandRegistry.cs

```diff
@@ -12,76 +12,66 @@ public abstract class CommandRegistry<TService> : Singleton<TService>,
+
+        public event ConnectedEventHandler Connected;
+        public event DisconnectedEventHandler Disconnected;
 
         protected void NotifyConnected(SocketEndpoint socketEndpoint)
         {
-            foreach (ICommandHandler commandHandler in GetCommandHandlerListForEnumeration())
-            {
-                commandHandler.OnConnected(socketEndpoint);
-            }
+            Connected?.Invoke(socketEndpoint);
         }
 
         protected void NotifyDisconnected(SocketEndpoint socketEndpoint)
         {
-            foreach (ICommandHandler commandHandler in GetCommandHandlerListForEnumeration())
-            {
-                commandHandler.OnDisconnected(socketEndpoint);
-            }
+            Disconnected?.Invoke(socketEndpoint);
         }
 
         protected void NotifyCommand(SocketEndpoint socketEndpoint, string command, BinaryReader message, int remainingDataSize)
         {
-            List<ICommandHandler> commandHandlerList = null;
             lock (lockObject)
             {
-                HashSet<ICommandHandler> commandHandlerSet;
-                if (commandHandlers.TryGetValue(command, out commandHandlerSet))
-                {
-                    commandHandlerList = new List<ICommandHandler>(commandHandlerSet);
-                }
-            }
-
-            if (commandHandlerList != null)
-            {
-                foreach (ICommandHandler commandHandler in commandHandlerList)
+                CommandHandler commandHandlerList = null;
+                if (commandHandlers.TryGetValue(command, out commandHandlerList))
                 {
-                    commandHandler.HandleCommand(socketEndpoint, command, message, remainingDataSize);
+                    commandHandlerList(socketEndpoint, command, message, remainingDataSize);
                 }
             }
         }
 
-        private HashSet<ICommandHandler> GetCommandHandlerListForEnumeration()
+        public void RegisterCommandHandler(string command, CommandHandler handler)
         {
             lock (lockObject)
             {
-                return new HashSet<ICommandHandler>(commandHandlers.Values.SelectMany(s => s));
-            }
-        }
-
-        public void RegisterCommandHandler(string command, ICommandHandler handler)
-        {
-            lock (lockObject)
-            {
-                HashSet<ICommandHandler> set;
-                if (!commandHandlers.TryGetValue(command, out set))
+                CommandHandler commandList;
+                if (!commandHandlers.TryGetValue(command, out commandList))
                 {
-                    commandHandlers[command] = set = new HashSet<ICommandHandler>();
+                    commandHandlers[command] = handler;
+                }
+                else
+                {
+                    commandHandlers[command] = commandList + handler;
                 }
```

Comment:
I think that we should throw if the specific command already has a handler. Handing the same BinaryReader to different handlers doesn't seem safe.

Reply:
I think that makes sense. I'll have to refactor one place where I'm intentionally handling the "Camera" command in two locations, one to actually read the values and the other to just tick an lastUpdatedTime for connectivity-monitoring purposes